### ISSUE TITLE
Automatic creation of launchReactNative.js and the debugger doesn't seem to take into account the `projectRoot` setting

### ISSUE THREAD

**Issue by Jarrio** [Microsoft/vscode-react-native#561]
### Missing launchReactNative.js
1. Looking through the issues, (I believe) this should be automatically generated if it has detected a react-native project.
2. Opened my project with react-native stuff in a `./bin` directory and my `./.vscode` directory at root and no file has been created, presumably because no `package.json` file is in my root.

### Debugger Issue
1. Make the `ProjectRoot` workspace setting a sub-folder
2. Press the debug button, debuggerWorker not found. My output is [below](#debugger-output)

Specifically, I have my output directory in a sub-folder called `bin` so if the path was `Crashdex\haxe\bin` then I would expect that to work. I'm not sure why it's looking for the worker in the `.vscode\.react` directory, isn't this an NPM module?

### LaunchReactNative.js
1. Open project and set a sub-folder as the react-native root
2. The extension to create the necessary file in the root `./.vscode` folder taking into account the variable

### Debugger
1. Open the debugger tab
2. Press "Play" and it should connect to the packager

## Software versions
- React-Native VS Code extension version: `0.4.4`
- VSCode version: `1.17.2`
- OS platform and version: Windows `10 64-bit`
- NodeJS version: `8.7.0`
- React Native version: `0.48.0`

# Debugger Output
```
OS: win32 x64
Adapter node: v7.9.0 x64
vscode-chrome-debug-core: 3.17.3
Starting debugger app worker.
Could not debug. ENOENT: no such file or directory, open 'w:\HaxeReact\Crashdex\haxe\.vscode\.react\debuggerWorker.js'
Error processing "disconnect": TypeError: Cannot read property 'noDebug' of undefined
    at Object.terminateSession (C:\Users\Jazz\.vscode\extensions\ms-vscode.node-debug2-1.18.3\node_modules\vscode-chrome-debug-core\out\src\chrome\chromeDebugAdapter.js:198:63)
    at Object.terminateSession (C:\Users\Jazz\.vscode\extensions\ms-vscode.node-debug2-1.18.3\out\src\nodeDebugAdapter.js:400:15)
    at Object.disconnect (C:\Users\Jazz\.vscode\extensions\ms-vscode.node-debug2-1.18.3\node_modules\vscode-chrome-debug-core\out\src\chrome\chromeDebugAdapter.js:821:21)
    at Object.dispatchRequest (C:\Users\Jazz\.vscode\extensions\vsmobile.vscode-react-native-0.4.4\node_modules\vscode-chrome-debug-core\out\src\chrome\chromeDebugSession.js:75:82)
    at Object.disconnect (C:\Users\Jazz\.vscode\extensions\vsmobile.vscode-react-native-0.4.4\src\debugger\nodeDebugWrapper.js:83:23)
    at Object.dispatchRequest (C:\Users\Jazz\.vscode\extensions\vsmobile.vscode-react-native-0.4.4\src\debugger\nodeDebugWrapper.js:44:29)
    at Object.ProtocolServer._handleData (C:\Users\Jazz\.vscode\extensions\vsmobile.vscode-react-native-0.4.4\node_modules\vscode-debugadapter\lib\protocol.js:104:38)
    at Socket.<anonymous> (C:\Users\Jazz\.vscode\extensions\vsmobile.vscode-react-native-0.4.4\node_modules\vscode-debugadapter\lib\protocol.js:24:60)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:191:7)
```
- Output of the React-Native output channel (View -> Toggle Output -> Select React-Native in ListBox):
```
[Info] Starting Packager

Scanning 568 folders for symlinks in w:\HaxeReact\Crashdex\haxe\bin\node_modules (15ms)
 ┌────────────────────────────────────────────────────────────────────────────┐ 
 │  Running packager on port 8081.                                            │ 
 │                                                                            │ 
 │  Keep this packager running while developing on any JS projects. Feel      │ 
 │  free to close this tab and run your own packager instance if you          │ 
 │  prefer.                                                                   │ 
 │                                                                            │ 
 │  https://github.com/facebook/react-native                                  │ 
 │                                                                            │ 
 └────────────────────────────────────────────────────────────────────────────┘ 
Looking for JS files in
   w:\HaxeReact\Crashdex\haxe\bin 


React packager ready.

[Info] Packager started.

Loading dependency graph, done.
Starting custom debugger by executing: echo A debugger is not needed:  "w:\HaxeReact\Crashdex\haxe\bin"

```
- Output of the Developer Tools console (Help -> Toggle Developer Tools -> Select Console tab):
```
Overwriting extension c:\Program Files\Microsoft VS Code\resources\app\extensions\ms-vscode.node-debug2 with C:\Users\Jazz\.vscode\extensions\ms-vscode.node-debug2-1.18.3.
e._logMessageInConsole @ extensionService.ts:435
keybindingsRegistry.ts:197 Ctrl+Alt+ keybindings should not be used by default under Windows. Offender:  e  for  java.projectConfiguration.update
e._assertNoCtrlAlt @ keybindingsRegistry.ts:197
mime.ts:59 Overwriting extension <<.js>> to now point to mime <<text/x-javascript>>
(anonymous) @ mime.ts:59
mime.ts:59 Overwriting extension <<.jsx>> to now point to mime <<text/x-javascript>>
(anonymous) @ mime.ts:59
mime.ts:59 Overwriting extension <<.es6>> to now point to mime <<text/x-javascript>>
(anonymous) @ mime.ts:59
2mime.ts:59 Overwriting extension <<.js>> to now point to mime <<text/x-javascriptreact>>
(anonymous) @ mime.ts:59
mime.ts:59 Overwriting extension <<.jsx>> to now point to mime <<text/x-javascriptreact>>
(anonymous) @ mime.ts:59
mime.ts:59 Overwriting extension <<.babel>> to now point to mime <<text/x-javascriptreact>>
(anonymous) @ mime.ts:59
2mime.ts:59 Overwriting extension <<.es6>> to now point to mime <<text/x-javascriptreact>>
(anonymous) @ mime.ts:59
console.ts:123 [Extension Host] (node:7664) DeprecationWarning: os.tmpDir() is deprecated. Use os.tmpdir() instead.
t.log @ console.ts:123
```

I would also like to mention my directory structure as it's probably a bit unconventional:
![image](https://user-images.githubusercontent.com/748557/31919984-e752be58-b85d-11e7-9d52-e416ac94819a.png)

I write my code in a language called Haxe, it's a strictly typed language that can compile to many other platforms. We have a thing called "externs" which can let us plug native libraries from C++ to Javascript and use them with the Haxe language.

There is a `react` and `react-native` community who have basically exposed them to haxe for people to use. I write my haxe code in the `src` directory which is then exported to javascript into the `./bin/generated/main.js` file where the source map also is. My `node_modules` is also in the `./bin` directory. I'm not sure if the necessary paths and stuff are configurable from within this extension as the explanation for the `launch.json` doesn't cover the parameters or what the `launchReactNative.js` is meant to contain.

Apologies for the long write-up. Wasn't sure if the issues should be seperate or not as they are kind of related (or may well be a setup issue on my behalf!)

**Comment by svekl** [comment]
Hi @Jarrio, 
`launchReactNative.js` file is not needed, so it's ok, that it doesn't exist, it is just a placeholder to satisfy VS Code limitation and is not used.

Your issue is that `debuggerWorker.js` file is not downloaded from packager to `w:\HaxeReact\Crashdex\haxe\.vscode\.react\debuggerWorker.js`

But I tried to reproduce that today in many ways and had no success. May I, please, ask you to create some example minimum project, where this issue is reproduced?

**Comment by Jarrio** [comment]
I appreciate that! I have created a basic project and have pre-compiled for both android and ios as I wasn't sure which target you're on (the ios one might be missing some `console.log` statements as I forgot to re-compile it when I added them.

I assume that these logs should be re-routed to the debugger right if everything is working correctly?

I had to (https://github.com/Jarrio/Haxe-React-Native-blank-project)[create a repo] for it due to the whole project size.

**Comment by svekl** [comment]
Hi @Jarrio,
thank you for an example, took a look and realized that I was partially wrong. File from `program` parameter is not used, but it's path is used. So, the only change in your project that I need to do to make it work is to return back
`"program": "${workspaceRoot}/.vscode/launchReactNative.js",`
instead of 
"program": "${workspaceRoot}/bin/generated/android/main.js",
in "Debug Android" in `launch.json`

**Comment by Jarrio** [comment]
Ah that's got it connected and connected! What's the intended use of the debugger? As a test I threw some RawText in my Render and it errors in app but the debugger doesn't respond to it, is that intended and is just meant to print the output in the `LogCat` Output?

**Comment by svekl** [comment]
Hi @Jarrio,
debugger is intended to work fine, including breakpoints, watches, debug console and other stuff. But for debugger to work, all RN syntax must be correct, or packager won't be able to work with that. So, if you will just throw some random text it's not gonna work.

But playing with your project I found a bug that breaks debugging for Haxe project on Windows machine, breakpoints are not working in `.hx` files even with correct project. I created a PR with fix, so, it will be included in some release soon.

As I see, you have an experience in VS Code extensions authoring, so you can not to wait for release and test fixed debugger behavior from `master` branch, PR is already merged. If you are interested in testing - here is contribution guide which might help you with setup https://github.com/Microsoft/vscode-react-native/blob/master/CONTRIBUTING.md

Thanks again for your report and example project, it helped a lot!

**Comment by iTOYS** [comment]
Fixed #565

**Comment by Jarrio** [comment]
I'm not sure why but since updating to `0.5` I cannot get the packager to connect at all now, the following is in the output:

```
[Error] Failed to start the React Native packager

Stack: TypeError: Failed to start the React Native packager: Cannot read property 'message' of undefined
	at Function.getWrappedError (C:\Users\Jazz\.vscode\extensions\vsmobile.vscode-react-native-0.5.0\src\common\error\internalError.js:35:16)
	at Function.wrapError (C:\Users\Jazz\.vscode\extensions\vsmobile.vscode-react-native-0.5.0\src\common\error\errorHelper.js:18:44)
	at resultOfCode.done.reason (C:\Users\Jazz\.vscode\extensions\vsmobile.vscode-react-native-0.5.0\src\common\entryPointHandler.js:40:72)
	at _rejected (C:\Users\Jazz\.vscode\extensions\vsmobile.vscode-react-native-0.5.0\node_modules\q\q.js:844:24)
	at C:\Users\Jazz\.vscode\extensions\vsmobile.vscode-react-native-0.5.0\node_modules\q\q.js:870:30
	at Promise.when (C:\Users\Jazz\.vscode\extensions\vsmobile.vscode-react-native-0.5.0\node_modules\q\q.js:1122:31)
	at Promise.promise.promiseDispatch (C:\Users\Jazz\.vscode\extensions\vsmobile.vscode-react-native-0.5.0\node_modules\q\q.js:788:41)
	at C:\Users\Jazz\.vscode\extensions\vsmobile.vscode-react-native-0.5.0\node_modules\q\q.js:604:44
	at runSingle (C:\Users\Jazz\.vscode\extensions\vsmobile.vscode-react-native-0.5.0\node_modules\q\q.js:137:13)
	at flush (C:\Users\Jazz\.vscode\extensions\vsmobile.vscode-react-native-0.5.0\node_modules\q\q.js:125:13)
```

**Comment by svekl** [comment]
Checked 0.5 with your sample project, looks like this issue is related to other changes, merged fix, new version is released, now works fine with your sample project.

**Comment by Jarrio** [comment]
Every time I start the packager it says `[Error] Current workspace is not a React Native project.` and then starts up anyway and works fine for that purpose. I'm not sure what I'm doing wrong but I can't get the debugger to connect to vscode at all.

My `output` just shows the following when attempting to connect via the debugger
```
Starting custom debugger by executing: echo A debugger is not needed:  "w:\HaxeReact\Alarm\bin"
[Info] Packager is already running.

Starting custom debugger by executing: echo A debugger is not needed:  "w:\HaxeReact\Alarm\bin"
```
and the app just times out after a little bit. In the debug console I get 
```
OS: win32 x64
Adapter node: v7.9.0 x64
vscode-chrome-debug-core: 3.18.4
```
And the following from the developer tools:
![image](https://user-images.githubusercontent.com/748557/32380946-58dd2a2e-c0a9-11e7-92ce-9bd65f44b19d.png)

**Comment by svekl** [comment]
Hi @Jarrio,
tried to reproduce your issue on this repo https://github.com/Jarrio/Haxe-React-Native-blank-project, but for me debugging works fine, the only change I have is 
```
            "name": "Debug Android",
            "program": "${workspaceRoot}/.vscode/launchReactNative.js",
```
Can you, please, try to reproduce it on this project? And also, please, check your `settings.json` `projectRoot` parameter
```
 "react-native-tools": {
        "projectRoot": "./bin"
    }  
```

**Comment by Jarrio** [comment]
I actually experience the same issues on that project as well. I can start the packager but it still says "Current project isn't a React-Native Project", I can't run `React Native: Run on Android Device` via the vscode `F1` menu and putting the app in debug mode it just never connects to the debugger and eventually times out

**Comment by svekl** [comment]
Ok, I see, what you mean, thanks, I get an error on `React Native: Run on Android Device` too. Can you try not to use `React Native: Run on Android Device` for now, but just use `Debug Android` from Debug tab?

**Comment by Jarrio** [comment]
Oh hey! That worked in the sample project and didn't in my other. Turns out I had a slight mis-spelling for the parameter in my settings.json file

**Comment by svekl** [comment]
As regards to error, that you had with `React Native: Run on Android Device` command and attaching to packager, it will be fixed after merge of #587
Thanks for reporting.

**Comment by Jarrio** [comment]
I appreciate all the help, thanks!
